cpp/tests/experimental/louvain_primitives.cu

```diff
@@ -0,0 +1,179 @@
+                   [b_key   = std::get<0>(tmp).begin(),
+                    e_key   = std::get<0>(tmp).end(),
+                    b_value = std::get<1>(tmp).begin(),
+                    e_value = std::get<1>(tmp).end()] __device__(auto) {
+                     char comma[2] = {0, 0};
+                     printf("keys = ");
+                     for (auto p = b_key; p != e_key; ++p) {
+                       printf("%s%d", comma, *p);
+                       comma[0] = ',';
+                     }
+                     printf("\n");
+                     printf("values = ");
+                     comma[0] = (char)0;
+                     for (auto p = b_value; p != e_value; ++p) {
+                       printf("%s%g", comma, *p);
+                       comma[0] = ',';
+                     }
+                     printf("\n");
+                   });
+
+  if (graph_t::is_multi_gpu) {
+    cugraph::experimental::sort_and_shuffle_kv_pairs(
+      handle.get_comms(),
+      std::get<0>(tmp).begin(),
+      std::get<0>(tmp).end(),
+      std::get<1>(tmp).begin(),
+      [] __device__(auto) { return 0; },
+      handle.get_stream());
+  }
+
+  rmm::device_vector<thrust::tuple<vertex_t, weight_t>> output_v(num_verts);
+
+  // NOTE: these are populated and shuffled before call to update_by_delta_modularity
+  rmm::device_uvector<weight_t> old_cluster_sum_v(num_verts, handle.get_stream());
+  rmm::device_uvector<weight_t> src_vertex_weight_v(num_verts, handle.get_stream());
+  rmm::device_uvector<weight_t> src_cluster_weight_v(num_verts, handle.get_stream());
+
+  copy_v_transform_reduce_key_aggregated_out_nbr(
+    handle,
+    graph_view,
+    thrust::make_zip_iterator(thrust::make_tuple(
+      old_cluster_sum_v.begin(), src_vertex_weight_v.begin(), src_cluster_weight_v.begin())),
+    cluster_v.begin(),
+    std::get<0>(tmp).begin(),
+    std::get<0>(tmp).end(),
+    std::get<1>(tmp).begin(),
+    [total_edge_weight, resolution] __device__(
+      auto src, auto neighbor_cluster, auto new_cluster_sum, auto src_info, auto a_new) {
+      auto old_cluster_sum = thrust::get<0>(src_info);
+      auto k_k             = thrust::get<1>(src_info);
+      auto a_old           = thrust::get<2>(src_info);
+
+      weight_t delta_modularity = 2 * (((new_cluster_sum - old_cluster_sum) / total_edge_weight) -
+                                       resolution * (a_new * k_k - a_old * k_k + k_k * k_k) /
+                                         (total_edge_weight * total_edge_weight));
+
+      printf("compute delta_modularity (%d, %d, %g, %g, %g, %g, %g)\n",
+             src,
+             neighbor_cluster,
+             new_cluster_sum,
+             old_cluster_sum,
+             k_k,
+             a_old,
+             a_new.load());
+
+      return thrust::make_tuple(neighbor_cluster, delta_modularity);
+    },
+    [] __device__(auto p1, auto p2) {
+      if (thrust::get<1>(p1) < thrust::get<1>(p2))
+        return p1;
+      else
+        return p2;
```

Comment:
```
return thrust::get<1>(p1) < thrust::get<1>(p2) ? p1 : p2;
```

Here ```p1```'s type is ```thrust::detail::tuple_of_iterator_references<int &, float &, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>```

and ```p2```'s type is ```thrust::tuple<int, float, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>```

so auto deduction of the return type based on ```return p1``` and ```return p2``` leads to conflict. Combining the two return statements to one fixes this.

If you pull the latest updates from PR 1399 and you apply this change as well, the code should at least compile.

Reply:
Done